### PR TITLE
bench/rttanalysis: make expectation validation fast and sane

### LINKED ISSUE #71107: bench/rttanalysis: TestBenchmarkExpectation failed
bench/rttanalysis.TestBenchmarkExpectation [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3536767&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3536767&tab=artifacts#/) on release-21.2 @ [1e0887f59f117988b9e7a30ae0946d3561247823](https://github.com/cockroachdb/cockroach/commits/1e0887f59f117988b9e7a30ae0946d3561247823):

```
BenchmarkAlterTableAddCheckConstraint/alter_table_add_3_check_constraints
BenchmarkDropTable/drop_2_tables-16        	       1	 576383048 ns/op	        27.00 roundtrips
BenchmarkDropTable/drop_3_tables
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkDropDatabase_drop_database_2_tables917572378
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkDropTable_drop_3_tables791856722
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableAddCheckConstraint_alter_table_add_3_check_constraints781278541
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableUnsplit_alter_table_unsplit_at_3_values191990782
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkCreateRole/create_role_with_1_option-16           	       1	 528724748 ns/op	        17.00 roundtrips
BenchmarkCreateRole/create_role_with_2_options
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkCreateRole_create_role_with_2_options186135541
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkSystemDatabaseQueries_select_system.users_with_empty_database_Name075038697
    test_log_scope.go:80: use -show-logs to present logs inline
BenchmarkDropView/drop_2_views-16        	       1	 316954533 ns/op	        28.00 roundtrips
BenchmarkDropView/drop_3_views
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkAlterTableDropConstraint/alter_table_drop_2_check_constraints-16        	       1	 970332886 ns/op	        17.00 roundtrips
BenchmarkAlterTableDropConstraint/alter_table_drop_3_check_constraints
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkDropView_drop_3_views868034841
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_3_check_constraints631533040
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableConfigureZone_alter_table_configure_zone_ranges958476862
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkVirtualTableQueries/select_crdb_internal.invalid_objects_with_1_fk-16         	       1	  49760321 ns/op	         1.000 roundtrips
=== CONT  TestBenchmarkExpectation/BenchmarkTruncate
    validate_benchmark_data.go:150: 
        	Error Trace:	validate_benchmark_data.go:150
        	            				validate_benchmark_data.go:88
        	Error:      	Received unexpected error:
        	            	exit status 6
        	Test:       	TestBenchmarkExpectation/BenchmarkTruncate
        	Messages:   	failed to wait for command /go/src/github.com/cockroachdb/cockroach/artifacts/go-build2030725776/b1017/rttanalysis.test --test.run=^$ --test.bench=BenchmarkTruncate --test.benchtime=1x --test.v
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkAlterRole/alter_role_with_3_options-16        	       1	 523498396 ns/op	        21.00 roundtrips
    --- FAIL: TestBenchmarkExpectation/BenchmarkTruncate (17.11s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestBenchmarkExpectation PKG=./pkg/bench/rttanalysis TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-experience pawalt
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBenchmarkExpectation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### LINKED ISSUE #71347: bench/rttanalysis: TestBenchmarkExpectation failed
bench/rttanalysis.TestBenchmarkExpectation [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3558597&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3558597&tab=artifacts#/) on master @ [09ea52347ceba714a734111ab5637efa2e14c609](https://github.com/cockroachdb/cockroach/commits/09ea52347ceba714a734111ab5637efa2e14c609):

```
=== RUN   TestBenchmarkExpectation
=== CONT  TestBenchmarkExpectation
    validate_benchmark_data.go:73: see the --rewrite flag to re-run the benchmarks and adjust the expectations
--- FAIL: TestBenchmarkExpectation (55.66s)
=== RUN   TestBenchmarkExpectation/BenchmarkTruncate
    validate_benchmark_data.go:137: /go/src/github.com/cockroachdb/cockroach/artifacts/go-build1189598317/b963/rttanalysis.test --test.run=^$ --test.bench=BenchmarkTruncate --test.benchtime=1x --test.v
=== CONT  TestBenchmarkExpectation/BenchmarkTruncate
    validate_benchmark_data.go:150: 
        	Error Trace:	validate_benchmark_data.go:150
        	            				validate_benchmark_data.go:88
        	Error:      	Received unexpected error:
        	            	exit status 6
        	Test:       	TestBenchmarkExpectation/BenchmarkTruncate
        	Messages:   	failed to wait for command /go/src/github.com/cockroachdb/cockroach/artifacts/go-build1189598317/b963/rttanalysis.test --test.run=^$ --test.bench=BenchmarkTruncate --test.benchtime=1x --test.v
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkRevokeRole_revoke_2_roles212711911
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkORMQueries_pg_namespace677933465
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkGrantRole_grant_2_roles784577962
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkORMQueries_pg_namespace651220004
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkGrantRole_grant_2_roles233247852
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkORMQueries_pg_namespace345977590
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkORMQueries/pg_namespace-16                               	       1	   8108442 ns/op	         2.000 roundtrips
BenchmarkORMQueries/pg_attribute
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkORMQueries_pg_attribute966982109
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkRevokeRole_revoke_2_roles131616282
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkORMQueries_pg_attribute164158999
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkGrantRole/grant_2_roles-16            	       1	 188360348 ns/op	        20.00 roundtrips
    --- FAIL: TestBenchmarkExpectation/BenchmarkTruncate (23.54s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestBenchmarkExpectation PKG=./pkg/bench/rttanalysis TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #71107 bench/rttanalysis: TestBenchmarkExpectation failed [C-test-failure O-robot branch-release-21.2]
</p>
</details>
/cc @cockroachdb/sql-experience pawalt
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBenchmarkExpectation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### LINKED ISSUE #72068: bench/rttanalysis: BenchmarkAlterTableDropConstraint failed
bench/rttanalysis.BenchmarkAlterTableDropConstraint [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3646670&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3646670&tab=artifacts#/) on master @ [12afa456ab273fdb640d35ea3b5c676a4474e914](https://github.com/cockroachdb/cockroach/commits/12afa456ab273fdb640d35ea3b5c676a4474e914):

```
--- FAIL: BenchmarkAlterTableDropConstraint
--- FAIL: BenchmarkAlterTableDropConstraint/alter_table_drop_1_check_constraint-16
BenchmarkAlterTableDropConstraint/alter_table_drop_2_check_constraints
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_2_check_constraints650254326
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_2_check_constraints231885976
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_2_check_constraints731649303
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_2_check_constraints713362785
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkAlterTableDropConstraint/alter_table_drop_2_check_constraints-16        	       1	 482866833 ns/op	        17.00 roundtrips
BenchmarkAlterTableDropConstraint/alter_table_drop_3_check_constraints
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_3_check_constraints070117708
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_3_check_constraints815541342
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_3_check_constraints336321061
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableDropConstraint_alter_table_drop_3_check_constraints607071135
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkAlterTableDropConstraint/alter_table_drop_3_check_constraints-16        	       1	 346711374 ns/op	        18.00 roundtrips
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-experience pawalt
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*BenchmarkAlterTableDropConstraint.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### LINKED ISSUE #73066: bench/rttanalysis: BenchmarkAlterTableAddColumn failed
bench/rttanalysis.BenchmarkAlterTableAddColumn [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3779676&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3779676&tab=artifacts#/) on master @ [c482da4c021d2dc83bce002d066a5f090d5f1384](https://github.com/cockroachdb/cockroach/commits/c482da4c021d2dc83bce002d066a5f090d5f1384):

```
--- FAIL: BenchmarkAlterTableAddColumn
--- FAIL: BenchmarkAlterTableAddColumn/alter_table_add_2_columns-16
BenchmarkAlterTableAddColumn/alter_table_add_3_columns
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableAddColumn_alter_table_add_3_columns325048454
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableAddColumn_alter_table_add_3_columns3685047004
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableAddColumn_alter_table_add_3_columns3916717696
    test_log_scope.go:80: use -show-logs to present logs inline
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logBenchmarkAlterTableAddColumn_alter_table_add_3_columns338933945
    test_log_scope.go:80: use -show-logs to present logs inline
    rtt_analysis_bench.go:131: -- test log scope end --
    rtt_analysis_bench.go:74: -- test log scope end --
BenchmarkAlterTableAddColumn/alter_table_add_3_columns-16        	       1	 221154663 ns/op	        15.00 roundtrips
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-experience
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*BenchmarkAlterTableAddColumn.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by ajwerner**
This commit reworks the way expectation validation is performed in the
rttanalysis framework. Before this change, we relied on subprocesses to
execute the benchmarks and then parsed the results. This was brittle and
slow. It was also exposed to issues where we still could not parallelize
sub-tests of individual benchmark categories.

We can't rearrange the top-level benchmarks because we want continuity in
the test names. Instead, we introduce a registry pattern which will store
the benchmark cases and then the top-level benchmarks can just exec
themselves from the registry. This does induce some strings, but it seems
worth it. We assert that indeed the benchmark is registered.

This gives us the ability to eschew most (but sadly, not all) of the
subprocess complexity and still get cross-benchmark parallelism. The
subprocess code which remains is about making sure that metamorphic
constants are disabled. It does let us eliminate a bunch of complexity
related to test pattern matching; we now delegate to the go testing
framework to decide which benchmarks to run in the expectation test.
This symmetry is pleasant. It also means that the `--rewrite` flag is now
a bool flag.

The key insight here is that we can use a testing.T and some additional
state to stand in for the previously expected testing.B via an interface.
This works out pretty well.

Fixes #72068
Fixes #71347
Fixes #71107
Fixes #73066

Before:
```
PASS
ok  	github.com/cockroachdb/cockroach/pkg/bench/rttanalysis	86.593s
```

After:
```
PASS
ok  	github.com/cockroachdb/cockroach/pkg/bench/rttanalysis	10.620s
```

Release note: None

### CODE REVIEW

File: pkg/bench/rttanalysis/testdata/benchmark_expectations [hunk 1]
```diff
@@ -60,8 +60,13 @@ exp,benchmark
 2,ORMQueries/has_column_privilege_using_column_name
 2,ORMQueries/has_table_privilege_real_table
 2,ORMQueries/has_table_privilege_virtual_table
+3,ORMQueries/information_schema._pg_index_position
 2,ORMQueries/pg_attribute
 2,ORMQueries/pg_class
+5,ORMQueries/pg_is_other_temp_schema
+9,ORMQueries/pg_is_other_temp_schema_multiple_times
+2,ORMQueries/pg_my_temp_schema
+2,ORMQueries/pg_my_temp_schema_multiple_times
```
Comment (ajwerner) [inline pkg/bench/rttanalysis/testdata/benchmark_expectations:69]: somehow we had just omitted these 🤔

**Comment by ajwerner** [comment]
@stevendanna maybe I can snipe you into reviewing this given you typed up  https://github.com/cockroachdb/cockroach/pull/71939 and this, in some ways, extends the principles of that change (though is much larger).

File: pkg/bench/rttanalysis/validate_benchmark_data.go [hunk 2]
```diff
@@ -58,193 +60,152 @@ var (
...
-	benchmarkFilter, err := regexp.Compile((*rewriteFlag)[:slashIdx])
-	require.NoError(t, err)
-
-	var g errgroup.Group
-	resChan := make(chan []benchmarkResult)
-	run := func(b string) {
-		if !benchmarkFilter.MatchString(b) {
+// execTestSubprocess execs the testing binary with all the same flags in order
+// to run it without metamorphic testing enabled. Metamorphic testing messes
+// with the benchmark results.
```
Comment (otan) [inline pkg/bench/rttanalysis/validate_benchmark_data.go:157]: so are we just running this to pass some tests? is that strictly necessary?
i'm confused why we don't just ignore metamorphic builds

File: pkg/bench/rttanalysis/validate_benchmark_data.go [hunk 3]
```diff
@@ -58,193 +60,152 @@ var (
...
-	benchmarks := getBenchmarks(t)
-	if *rewriteFlag != "" {
-		rewriteBenchmarkExpectations(t, benchmarks)
+func runBenchmarkExpectationTests(t *testing.T, r *Registry) {
+	if util.IsMetamorphicBuild() {
+		execTestSubprocess(t)
 		return
 	}
 
+	defer log.Scope(t).Close(t)
```
Comment (otan) [inline pkg/bench/rttanalysis/validate_benchmark_data.go:69]: i wonder if this belongs in `TestBenchmarkExpectations` at the top level, but not fussed. i  guess would it be worth including `//go:generate ../../util/leaktest/add-leaktest.sh *_test.go` somewhere would have automatically added this.

File: pkg/bench/rttanalysis/validate_benchmark_data.go [hunk 4]
```diff
@@ -58,193 +60,152 @@ var (
...
+	var results resultSet
+	var wg sync.WaitGroup
+	concurrency := system.NumCPU() * 4 // arbitrary
+	limiter := quotapool.NewIntPool("rttanalysis", uint64(concurrency))
+	isRewrite := *rewriteFlag
+	for b, cases := range r.r {
+		quota, err := limiter.Acquire(context.Background(), 1)
+		require.NoError(t, err)
+		wg.Add(1)
+		go func(b string, cases []RoundTripBenchTestCase) {
```
Comment (otan) [inline pkg/bench/rttanalysis/validate_benchmark_data.go:86]: ya reckon this will be ok in parallel if multiregion tests are starting 9 servers each in CI?

**Review by otan — COMMENTED** [review]
cursory glance, looks nicer to use!

File: pkg/bench/rttanalysis/validate_benchmark_data.go [hunk 5]
```diff
@@ -58,193 +60,152 @@ var (
...
+		vs := f.Value.String()
+		switch f.Name {
+		case "run", "test.run":
+			prefix := "^" + regexp.QuoteMeta(t.Name()) + "$"
+			if idx := strings.Index(vs, "/"); idx >= 0 {
+				vs = prefix + vs[idx:]
+			} else {
+				vs = prefix
+			}
+		case "bench", "test.bench":
```
Comment (stevendanna) [inline pkg/bench/rttanalysis/validate_benchmark_data.go:170]: Is this flag mangling and the addition of `--test.bench=^$` below still necessary?  If so might be good to leave ourselves a comment about the case this is handling.

**Review by stevendanna — COMMENTED** [review]
I've tried this out locally and it works pretty great.  We can now select tests down to the sub test without issue and the speedup is pretty nice locally as well.

It's a little strange having to define your tests in init(). Seems worth the tradeoff to me, but I don't work on these tests.

Thanks for digging into this. Hopefully this improves the reliability of these in CI and makes it easier to locally stresstest any flakes.

**Review by ajwerner — COMMENTED** [review]
> It's a little strange having to define your tests in init(). Seems worth the tradeoff to me, but I don't work on these tests.

Yeah, I wavered for a while on what to actually do here and ended up being pretty happy with this. From a usability perspective, adding a new test hasn't really changed. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-:-MpHYkC15PKGtIyGZViE:b-kzi33r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @stevendanna)

___
*[pkg/bench/rttanalysis/alter_table_bench_test.go, line 16 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFIlbbUgm5l0xec-r2-16:-MpHWpHp0Pl1U83fIujh:b-q6a20q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/alter_table_bench_test.go#L16)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit / not important: any reason this isn't in one big `init()` clause?

same as below tests, not too fussed either way
</blockquote></details>

It was to make the relationship between the cases and the top-level benchmark more obvious. Also to make the diff smaller. I'm phobic of breaking existing benchmark names. 
___
*[pkg/bench/rttanalysis/cluster.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFRDDurbByiFUkZ-r2-50:-MpHX-VK2wNsYLEwMQ1A:b-5yzl4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/cluster.go#L50)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

super nit: `stmtToKvBatchRequests`
</blockquote></details>

Are you saying you want `KV`? I think you typed it as it is. Changed it to `KV`. 
___
*[pkg/bench/rttanalysis/testing_shims.go, line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFRDDurbByiFUka-r2-51:-MpHX7RW3t_qqnsQ-3G0:b-sbjviv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/testing_shims.go#L51)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

is this testing.T or testing.B? bShim looks like a testing.B
</blockquote></details>

Well, the `tb` interface is more of a `B` than a `T`, but it's really nice to be able to pass `tShim` here to the test as it allows for parallelism. I've cleaned up the language and changed the interface to `testingB`. We're faking the `testing.B` methods on the `testing.T` and recording the metric output. 
___
*[pkg/bench/rttanalysis/validate_benchmark_data.go, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFIlbbUgm5l0xem-r2-69:-MpHXfP462QMZ1wFJHRd:bh3t93h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/validate_benchmark_data.go#L69)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i wonder if this belongs in `TestBenchmarkExpectations` at the top level, but not fussed. i  guess would it be worth including `//go:generate ../../util/leaktest/add-leaktest.sh *_test.go` somewhere would have automatically added this.
</blockquote></details>

It was irritating to see a log scope get created before the subprocess exec so I pulled it down here. Commented.
___
*[pkg/bench/rttanalysis/validate_benchmark_data.go, line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFIlbbUgm5l0xem-r2-86:-MpHXpg_4PK3g27EAVHH:b-umic22) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/validate_benchmark_data.go#L86)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

ya reckon this will be ok in parallel if multiregion tests are starting 9 servers each in CI?
</blockquote></details>

That's legit, added a thing to divide the parallelism by the number of servers per test. 
___
*[pkg/bench/rttanalysis/validate_benchmark_data.go, line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFIlbbUgm5l0xem-r2-157:-MpHY4NnAt2s2dMXAO5r:byjjlu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/validate_benchmark_data.go#L157)):*
> i'm confused why we don't just ignore metamorphic builds

Well, in CI all of our unit tests are metamorphic builds. Ignoring them would make this test not run in CI (which is actually what happened for several months :frowning_face:). 
___
*[pkg/bench/rttanalysis/validate_benchmark_data.go, line 170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFIlbbUgm5l0xem-r2-170:-MpHYIIn4KpgU4Xqhkjf:b-ndwx6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/validate_benchmark_data.go#L170)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Is this flag mangling and the addition of `--test.bench=^$` below still necessary?  If so might be good to leave ourselves a comment about the case this is handling.
</blockquote></details>

Yes, consider the case where to user runs `go test --run Expectations --bench .' which would run all the tests and all of the benchmarks. If we didn't exclude the benchmarks in the subprocess, we'd run them twice. 
___
*[pkg/bench/rttanalysis/validate_benchmark_data.go, line 240 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73097#-MpEdNFIlbbUgm5l0xem-r2-240:-MpHfM1xA3NfKsRZZ6J3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a6721a635410b807676e1a6391e92f712e149d/pkg/bench/rttanalysis/validate_benchmark_data.go#L240)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] s/expecatations/expectations/
</blockquote></details>

Done.

**Review by stevendanna — APPROVED** [review]

**Comment by yuzefovich** [comment]
Thanks for improving this!

I'm still a bit confused though - how does one rewrite the expectations? I tried running `make bench PKG=./pkg/bench/rttanalysis BENCHES=Revoke TESTFLAGS='-rewrite'`, and although the run "passed", it indicated different numbers for some subtests and didn't rewrite the expectations file. I think it should be reproducible if you pull #68430 and revert the changes to the expectations file.

**Comment by ajwerner** [comment]
You don't run the benches to rewrite, you run the test. `make test PKG=./pkg/bench/rttanalysis TESTS=Expectation/Revoke TESTFLAGS='-rewrite'`

**Pull request merged**